Reorganize tests

aelanman opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

A massive overhaul of the unit tests, including a variety of other fixes:
- `create_mock_catalog` can now make diffuse models via the optional `analytic_diffuse` module.
- These diffuse models have analytic solutions, and a new unit test checks pyuvsim results against the analytic solution for several models.
- Switches to using tmpdir instead of custom-made temporary directories.
- Tests that visibilities for a zenith source with a power law spectrum follow the same spectrum.
- Has analytic (point source) visibility tests use multiple beam models. This revealed an error in the ordering of visibility components in the test, which is now fixed.
- Increases frequency samples of the saved UVBeam model, so that cubic interpolation (default) may be used in tests.
- Switches to using a new Counter class that doesn't rely on running a thread on the root process. This has better performance overall (a future PR adding MPI-enabled tests also shows this).
- Separates `test_mpi_uvsim.py` into `test_mpi.py` and `test_runs.py`. The first is for tests of various mpi-related functions (which will be parallelized in a future PR). The second is meant for tests of full simulation runs.
- Found and fixed a bug related to options given to analytic beams.
- Finally fixes the hacky "test_z_profiler.py" naming that was done to ensure profiler tests ran last. The tests are strictly ordered by using pytest hooks.

**Important** This is not compatible with earlier versions of pyradiosky that had non-Quantity stokes arrays. For that reason, this may warrant raising the minor version number on pyuvsim, and making a new pyradiosky release.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

The unit tests are long overdue for review, and were in need of some upgrades to better take advantage of pytest features. These changes clarify the purposes of several tests, improve organization, and cover more cases.

<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->
Closes #274 
Closes #302 

## Types of changes
<!--- What types of changes does your code introduce? Put an replace the space with an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Reference simulation update or replacement
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

Bug fix checklist:
- [x] My fix includes a new test that breaks as a result of the bug (if possible).
- [x] All new and existing tests pass.
- [ ] I have checked that I reproduce the reference simulations or if there are differences they are explained below (if appropriate). If there are changes that are correct, I will update the reference simulation files after this PR is merged.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvsim/blob/master/CHANGELOG.md).

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the documentation to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [ ] I have checked that I reproduce the reference simulations or if there are differences they are explained below (if appropriate). If there are changes that are correct, I will update the reference simulation files after this PR is merged.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvsim/blob/master/CHANGELOG.md).

Breaking change checklist:
- [x] I have updated the docstrings associated with my change using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] I have updated the documentation to reflect my changes (if appropriate).
- [x] My change includes backwards compatibility and deprecation warnings (if possible).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass.
- [ ] I have checked that I reproduce the reference simulations or if there are differences they are explained below (if appropriate). If there are changes that are correct, I will update the reference simulation files after this PR is merged.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvsim/blob/master/CHANGELOG.md).

aelanman commented on ci/tests.yaml:
```diff
@@ -19,5 +19,6 @@ dependencies:
   - setuptools_scm
   - pip:
-    - pyradiosky
+    - git+git://github.com/RadioAstronomySoftwareGroup/pyradiosky
```
These changes are not compatible with the current release, because they rely on SkyModel.stokes array being of Quantity type. I figure we'd wait to merge this until a new release of pyradiosky, and then switch it back to using the released version before merging.

mkolopanis replied:
Oh okay that makes a lot of sense. Let's make an issue so we don't forget to go back to the released version when able.

bhazelton replied:
We cut a new pyradiosky version (v0.1.0) last week and there are no new commits after that label, so I think you should be able to use the PyPI package.

aelanman commented on pyuvsim/tests/test_run.py:
```diff
@@ -0,0 +1,233 @@
+    assert np.allclose(ana / 2, dat, atol=1e-2)
+
+
+@pytest.mark.filterwarnings("ignore:The frequency field is included in the recarray")
+def test_run_paramdict_uvsim():
+    # Running a simulation from parameter dictionary.
+
+    params = pyuvsim.simsetup._config_str_to_dict(
+        os.path.join(SIM_DATA_PATH, 'test_config', 'param_1time_1src_testcat.yaml')
+    )
+
+    pyuvsim.run_uvsim(params, return_uv=True)
```
This was an old test, so I can't speak to the original intent, but that seems to be the case.

mkolopanis replied:
Hm interesting. I'll make an issue to investigate further to see if there is something more concrete we can check.

aelanman commented on pyuvsim/tests/test_simsetup.py:
```diff
@@ -77,22 +73,51 @@ def test_mock_catalog_off_zenith_source():
 
 
+def test_mock_diffuse_maps():
+    # TODO -- Also run for moonlocation
```
Ah, I should add that to this PR. Thanks for spotting it.

aelanman commented:
@mkolopanis Thanks for looking this over! I'll make the requested changes. I can try atomizing the tests further as well, but I'm also happy to leave that to a future PR. This was more of a "first pass" at fixing up the tests.

aelanman commented:
@bhazelton @mkolopanis  I think I've addressed all your comments, and restored the coverage that was lost. I noticed some lines in mpi.py that should end up covered when the mpi-enabled tests are merged. Currently, PR #310 is set to merge into this branch.

I guess the things to decide now are:
1. Do we want to address issue #311 here, or in a future change?
2. Likewise, further atomizing tests? (It might help to identify a few tests that could be better split up).
3.  Should #310  be merged into here, or this merged into master first?

bhazelton commented:
I'm fine with merging this before #310 and waiting to fix #311 and the atomization. If we wait on the atomization, though, we should make an issue for it.

aelanman commented:
@bhazelton I think this might also deserve a minor version bump, since it loses support for older versions of pyradiosky?

mkolopanis commented on pyuvsim/tests/test_antenna.py:
```diff
@@ -44,9 +43,38 @@ def test_get_beam_jones():
     vers = pyuvdata.__version__.split('.')
     version = (float(vers[0]), float(vers[1]), float(vers[2]))
-    if version[0] < 2 or (version[0] >= 2 and version[2] < 1):
+    if version[0] < 2 or (version[0] >= 2 and version[2] < 1 and version[1] < 1):
```
Won't this line trigger if `version = (3,0,0)`? Should the first part inside the parentheses just be `version[0] == 2`

aelanman replied:
I think you're right... this should be checking for the case that the version is < 2.0.0. Since the tests are running the latest version of pyuvdata, though, we might as well drop this altogether.
